src/lib/tls/tls_session.cpp

```diff
@@ -138,39 +175,26 @@ Session_Summary::Session_Summary(const Server_Hello_13& server_hello,
 
    // In TLS 1.3 the key exchange algorithm is not negotiated in the ciphersuite
    // anymore. This provides a compatible identifier for applications to use.
-   m_kex_algo = kex_method_to_string([&] {
+
+   const auto group = [&]() -> std::optional<Named_Group> {
       if(psk_used() || was_resumption()) {
          if(const auto keyshare = server_hello.extensions().get<Key_Share>()) {
-            const auto group = keyshare->selected_group();
-            if(group.is_dh_named_group()) {
-               return Kex_Algo::DHE_PSK;
-            } else if(group.is_ecdh_named_curve() || group.is_x25519() || group.is_x448()) {
-               return Kex_Algo::ECDHE_PSK;
-            } else if(group.is_pure_kyber()) {
-               return Kex_Algo::KEM_PSK;
-            } else if(group.is_pqc_hybrid()) {
-               return Kex_Algo::HYBRID_PSK;
-            }
+            return keyshare->selected_group();
          } else {
-            return Kex_Algo::PSK;
+            return {};
          }
       } else {
          const auto keyshare = server_hello.extensions().get<Key_Share>();
          BOTAN_ASSERT_NONNULL(keyshare);
-         const auto group = keyshare->selected_group();
-         if(group.is_dh_named_group()) {
-            return Kex_Algo::DH;
-         } else if(group.is_ecdh_named_curve() || group.is_x25519() || group.is_x448()) {
-            return Kex_Algo::ECDH;
-         } else if(group.is_pure_kyber()) {
-            return Kex_Algo::KEM;
-         } else if(group.is_pqc_hybrid()) {
-            return Kex_Algo::HYBRID;
-         }
+         return keyshare->selected_group();
       }
+   }();
+
+   if(group.has_value()) {
+      m_kex_parameters = group->to_string();
+   }
```

Comment:
C++23 will bring [monadic stuff to `std::optional<>`](https://en.cppreference.com/w/cpp/utility/optional/transform) finally. Along those lines:

```C++
m_kex_parameters = group.transform([](auto& x) { return x.to_string(); });
```

... the bulky C++ lambdas somewhat spoil the fun; still better than three lines of `if`, I guess.

Reply:
I'm definitely looking forward to it, I miss Rust's monadic operations on `std::optional` a lot of the time working in C++

As for pattern matching (sigh)